sway-lib-std/src/ufp64.sw

```diff
@@ -0,0 +1,247 @@
+library ufp64;
+//! A wrapper around u64 type for a library for Sway for mathematical functions operating with signed 32.32-bit fixed point numbers. 
+
+use core::num::*;
+use ::assert::assert;
+use ::math::*;
+use ::logging::*;
+use ::revert::revert;
+use ::u128::U128;
+
+pub struct UFP64 {
+    value: u64 
+}
+
+impl UFP64 {
+    pub fn denominator() -> u64 {
+        1 << 32
+    }
+
+    pub fn zero() -> UFP64 {
+        UFP64 {
+            value: 0,
+        }
+    }
+
+    /// The smallest value that can be represented by this type.
+    pub fn min() -> UFP64 {
+        UFP64 {
+            value: ~u64::min(),
+        }
+    }
+
+    /// The largest value that can be represented by this type,
+    pub fn max() -> UFP64 {
+        UFP64 {
+            value: ~u64::max(),
+        }
+    }
+
+    /// The size of this type in bits.
+    pub fn bits() -> u32 {
+        64
+    }
+}
+
+impl core::ops::Eq for UFP64 {
+    pub fn eq(self, other: Self) -> bool {
+        self.value == other.value
+    }
+}
+
+impl core::ops::Ord for UFP64 {
+    pub fn gt(self, other: Self) -> bool {
+        self.value > other.value
+    }
+
+    pub fn lt(self, other: Self) -> bool {
+        self.value < other.value
+    }
+}
+
+impl core::ops::Add for UFP64 {
+    /// Add a UFP64 to a UFP64. Panics on overflow.
+    pub fn add(self, other: Self) -> Self {
+        UFP64 {
+            value: self.value + other.value
+        }
+    }
+}
+
+impl core::ops::Subtract for UFP64 {
+    /// Subtract a UFP64 from a UFP64. Panics of overflow.
+    pub fn subtract(self, other: Self) -> Self {
+        // If trying to subtract a larger number, panic.
+        assert(!(self.value < other.value));
+
+        UFP64 {
+            value: self.value - other.value
+        }
+    }
+}
+
+impl core::ops::Multiply for UFP64 {
+    /// Multiply a UFP64 with a UFP64. Panics of overflow.
+    pub fn multiply(self, other: Self) -> Self {
+
+        let s_value = ~U128::from(0, self.value);
+        let o_value = ~U128::from(0, other.value);
+
+        let s_mul_o = s_value * o_value;
+        let res_u128 = s_mul_o >> 32;
+        if res_u128.upper != 0 {
+            // panic on overflow
```

Comment:
Wouldn't this panic on its own already? Is this needed cc @adlerjohn @nfurfaro ?

Reply:
No, it won't as the operation is done in u128, not u64